File: Stack/Min Stack/MinMain.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

type Pair struct {
	val      int
	minValue int
}

type MinStack struct {
	stack []Pair
}

func Constructor() *MinStack {
	return &MinStack{stack: []Pair{}}
}

func (ms *MinStack) Push(val int) {
	if len(ms.stack) == 0 {
		ms.stack = append(ms.stack, Pair{val: val, minValue: val})
	} else {
		currentMin := ms.stack[len(ms.stack)-1].minValue
		ms.stack = append(ms.stack, Pair{val: val, minValue: min(val, currentMin)})
	}
}

func (ms *MinStack) Pop() {
	if len(ms.stack) > 0 {
		ms.stack = ms.stack[:len(ms.stack)-1]
	}
}

func (ms *MinStack) Top() int {
	if len(ms.stack) > 0 {
		return ms.stack[len(ms.stack)-1].val
	}
	panic("Stack is empty")
}

func (ms *MinStack) GetMin() int {
	if len(ms.stack) > 0 {
		return ms.stack[len(ms.stack)-1].minValue
	}
	panic("Stack is empty")
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func main() {
	scanner := bufio.NewScanner(os.Stdin)
	scanner.Scan()
	operationsCount, _ := strconv.Atoi(scanner.Text())
	minStack := Constructor()
	for i := 0; i < operationsCount; i++ {
		fmt.Println("Enter operation (push <value> / pop / top / getMin):")
		scanner.Scan()
		command := strings.Split(scanner.Text(), " ")
		switch command[0] {
		case "push":
			val, _ := strconv.Atoi(command[1])
			minStack.Push(val)
			fmt.Println("Pushed:", val)
		case "pop":
			if len(minStack.stack) > 0 {
				minStack.Pop()
				fmt.Println("Popped top element")
			} else {
				fmt.Println("Stack is empty")
			}
		case "top":
			func() {
				defer func() {
					if r := recover(); r != nil {
						fmt.Println(r)
					}
				}()
				fmt.Println("Top element:", minStack.Top())
			}()
		case "getMin":
			func() {
				defer func() {
					if r := recover(); r != nil {
						fmt.Println(r)
					}
				}()
				fmt.Println("Minimum element:", minStack.GetMin())
			}()
		default:
			fmt.Println("Invalid command")
		}
	}
}

File: Stack/Min Stack/MinMain.cpp
#include <iostream>
#include <sstream>
#include <stack>
#include <stdexcept>
#include <string>
using namespace std;

class MinStack {
 private:
  struct Pair {
    int val, minValue;
    Pair(int val, int minValue) : val(val), minValue(minValue) {}
  };
  stack<Pair> st;

 public:
  MinStack() {}

  void push(int val) {
    if (st.empty()) {
      st.push(Pair(val, val));
    } else {
      int currentMin = min(st.top().minValue, val);
      st.push(Pair(val, currentMin));
    }
  }

  void pop() {
    if (!st.empty()) {
      st.pop();
    } else {
      throw runtime_error("Stack is empty");
    }
  }

  int top() {
    if (!st.empty()) {
      return st.top().val;
    }
    throw runtime_error("Stack is empty");
  }

  int getMin() {
    if (!st.empty()) {
      return st.top().minValue;
    }
    throw runtime_error("Stack is empty");
  }
};

int main() {
  MinStack minStack;
  cout << "Enter number of operations:" << endl;
  int operationsCount;
  cin >> operationsCount;
  cin.ignore();
  for (int i = 0; i < operationsCount; i++) {
    cout << "Enter operation (push <value> / pop / top / getMin):" << endl;
    string input;
    getline(cin, input);
    istringstream iss(input);
    string command;
    iss >> command;
    try {
      if (command == "push") {
        int value;
        iss >> value;
        minStack.push(value);
        cout << "Pushed: " << value << endl;
      } else if (command == "pop") {
        minStack.pop();
        cout << "Popped top element" << endl;
      } else if (command == "top") {
        cout << "Top element: " << minStack.top() << endl;
      } else if (command == "getMin") {
        cout << "Minimum element: " << minStack.getMin() << endl;
      } else {
        cout << "Invalid command" << endl;
      }
    } catch (const exception& e) {
      cout << e.what() << endl;
    }
  }
  return 0;
}
